File: observe/hook.py
#!/usr/bin/env python3
"""Firth Observe — Claude Code PostToolUse hook (read-only).

Reads the PostToolUse event JSON from stdin, scans it for credential touches /
exposures, and appends redacted findings to ``$CLAUDE_PROJECT_DIR/.firth/audit.jsonl``.

Guarantees:
  * Always exits 0 and writes nothing to stdout -> never blocks or alters a tool.
  * Never persists raw secret values (see scanner.py fingerprinting).
  * Sends nothing off the machine.
"""

import datetime
import json
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from scanner import scan_event  # noqa: E402


def main():
    try:
        event = json.load(sys.stdin)
    except Exception:
        return 0

    base = os.environ.get("CLAUDE_PROJECT_DIR") or event.get("cwd") or "."
    firth_dir = os.path.join(base, ".firth")
    firth_abs = os.path.abspath(firth_dir)
    self_dir = os.path.dirname(os.path.abspath(__file__))

    def ignore(path):
        try:
            ap = os.path.abspath(path)
        except Exception:
            return False
        return ap.startswith(firth_abs) or ap.startswith(self_dir)

    try:
        findings = scan_event(event, ignore_path=ignore)
    except Exception as exc:  # never let the hook disrupt the session
        print("firth-observe: scan error: {}".format(exc), file=sys.stderr)
        return 0

    if not findings:
        return 0

    common = {
        "ts": datetime.datetime.now(datetime.timezone.utc).isoformat(),
        "session_id": event.get("session_id"),
        "tool": event.get("tool_name"),
        "cwd": event.get("cwd"),
    }
    try:
        os.makedirs(firth_dir, exist_ok=True)
        with open(os.path.join(firth_dir, "audit.jsonl"), "a", encoding="utf-8") as fh:
            for fd in findings:
                rec = dict(common)
                rec.update(fd)
                fh.write(json.dumps(rec, ensure_ascii=False) + "\n")
    except Exception as exc:
        print("firth-observe: write failed: {}".format(exc), file=sys.stderr)

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: cli/src/observe/hook.ts
import { appendFileSync, mkdirSync } from 'node:fs'
import { dirname, join, resolve } from 'node:path'
import { fileURLToPath, pathToFileURL } from 'node:url'
import { scanEvent, type ToolEvent } from './scanner.js'

export function recordFindings(event: ToolEvent, baseDir: string): number {
  const firthDir = join(baseDir, '.firth')
  const firthAbs = resolve(firthDir)
  const selfDir = dirname(fileURLToPath(import.meta.url))
  const ignore = (p: string): boolean => {
    let ap: string
    try { ap = resolve(p) } catch { return false }
    return ap.startsWith(firthAbs) || ap.startsWith(selfDir)
  }
  const findings = scanEvent(event, { ignorePath: ignore })
  if (findings.length === 0) return 0
  const common = {
    ts: new Date().toISOString(),
    session_id: event.session_id ?? null,
    tool: event.tool_name ?? null,
    cwd: event.cwd ?? null,
  }
  mkdirSync(firthDir, { recursive: true })
  const lines = findings.map((f) => JSON.stringify({ ...common, ...f })).join('\n') + '\n'
  appendFileSync(join(firthDir, 'audit.jsonl'), lines, 'utf8')
  return findings.length
}

async function readStdin(): Promise<string> {
  const chunks: Buffer[] = []
  for await (const c of process.stdin) chunks.push(c as Buffer)
  return Buffer.concat(chunks).toString('utf8')
}

export async function main(): Promise<void> {
  let event: ToolEvent
  try { event = JSON.parse(await readStdin()) } catch { process.exit(0) }
  const base = process.env.CLAUDE_PROJECT_DIR || (event.cwd as string) || '.'
  try { recordFindings(event, base) } catch (e) { process.stderr.write(`firth-observe: ${e instanceof Error ? e.message : e}\n`) }
  process.exit(0)
}

// Run only when executed directly (the materialized .firth/observe/hook.js entry).
// Imported by tests → guard is false → main() does not run.
if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) void main()
